### PR TITLE
8223717: javafx printing: Support Specifying Print to File in the API

### PR DESCRIPTION
**Pull request by prrace**
This enhancement adds the String property outputFileProperty() to the JobSettings class.
The value should be a string that references a local file encoded as a URL.
If this is non-null and set to a location that the user has permission to write to,
then the printer output will be spooled there instead of the printer, so long as the platform printing system supports this.
The user can of course also set a print-to-file destination in the platform printer dialogs which may over-ride what the application set. But now the application can also see what it was set to, and cancel or alter it if necessary.

A simple manual test is provided, manual mainly because the few real printing functional tests are all manual as they are only useful if run with a printer configured.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8223717](https://bugs.openjdk.java.net/browse/JDK-8223717): javafx printing: Support Specifying Print to File in the API

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Prasanta Sadhukhan](https://openjdk.java.net/census#psadhukhan) (@prsadhuk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/543/head:pull/543` \
`$ git checkout pull/543`

Update a local copy of the PR: \
`$ git checkout pull/543` \
`$ git pull https://git.openjdk.java.net/jfx pull/543/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 543`

View PR using the GUI difftool: \
`$ git pr show -t 543`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/543.diff">https://git.openjdk.java.net/jfx/pull/543.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 1]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+    ///////////////////////  START OUTPUTFILE /////////////////////
+
+    private static final String DEFAULT_OUTPUTFILE = "";
+    private SimpleStringProperty outputFile;
+
+    /**
+     * A {@code StringProperty} representing the {@code java.net.URL}
+     * encoded name of a filesystem file, to which the platform printer
+     * driver should spool the rendered print data.
+     * <p>
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: Can you add a sentence indicating that the URL needs to be a `file:` URL? And indicate what happens if it isn't (I presume it is ignored)?

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 2]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+     * Applications can use this to request print-to-file behaviour where
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:475]: behaviour -> behavior

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 3]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+     * This is often useful where the printer driver generates a format
+     * such as Postscript or PDF, and the application intends to distribute
+     * the result instead of printing it, or for some other reason the
+     * application does not want physical media (paper) emitted by the printer.
+     * <p>
+     * If the application does not have permission to write to the specified
+     * file, a {@code SecurityException} may be thrown when printing.
+     * If the print system does not support print-to-file, then this
+     * <p>
+     * setting will be ignored.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:488]: These two lines should be switched.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 4]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+     * If the URL specifies a non-existent path, or does not specify
+     * a writable file it may be ignored, or a SecurityException may be thrown.
+     * The default value is an empty string, which is interpreted as unset,
+     * which means output is sent to the printer.
+     *
+     * @return the name of a printer spool file
+     * @since 13
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java]: 17

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 5]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+     * the result instead of printing it, or for some other reason the
+     * application does not want physical media (paper) emitted by the printer.
+     * <p>
+     * If the application does not have permission to write to the specified
+     * file, a {@code SecurityException} may be thrown when printing.
+     * If the print system does not support print-to-file, then this
+     * <p>
+     * setting will be ignored.
+     * If the URL specifies a non-existent path, or does not specify
+     * a writable file it may be ignored, or a SecurityException may be thrown.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:490]: I want to look at what we do for other properties like this. I'm not sure how feasible it is to throw an exception back to the application. And I wouldn't want to single out a security exception (what about a read-only file system?). I think that whatever the reason, if the file can't be written to, we should treat it consistently. Probably by ignoring the request. I presume that's what will happen if the print driver doesn't support print-to-file?

**Review by kevinrushforth — COMMENTED** [review]
I left some preliminary comments on the API spec.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 6]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+        }
+        return outputFile;
+    }
+
+    /**
+     * Get the output file path name.
+     * @return the output file as a string encoded {@code java.net.URL}
+     * @since 17
+     */
+    public String getOutputFile() {
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:541]: This should be final (as should the setter). I see that other getter and setter methods in this class are also non-final, but we shouldn't propagate this mistake.

Also, you don't need javadoc comments on the getters and setters, since it will copy them from the property (and add appropriate links).

**Review by kevinrushforth — COMMENTED** [review]
One more comment on the API for the getters and setter methods.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 5 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:490]: Not easy to specify exactly. We are handing it off to 2D
We aren't doing any up-front verification here, but I don't want to copy 2D behaviour into the docs, even if I could be sure about that .. I may just want to be more waffly about what happens.
Could be printed to the printer, could be a failed job, could be an exception of some kind if some nice piece of Java code recognises a problem up-front

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 1 above)
Comment (swpalmer) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: As a potential user of this API I have to ask, why would this be a URL instead of a File or Path ?  Particularly if only "file:" URLs will be valid.  I can't think of a scenario where it won't always be an extra inconvenience to convert the path I have to a URL.

Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: We aren't ignoring it .. its just handed on, so its going to be the same as the other non-writable case .. file could not be created .. printing error .. 
we could do some up-front validation if that seems like the way we want to go. This could I guess also check if the file is writeable .. but since the user can select a non-writable file location in the dialog I don't think we should do that part. In which case I wonder about the URL validation too ..

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 6 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:541]: But the JobSettings class is final .. is it still necessary ?

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 1 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: Consistency with other APIs would be the main reason. We could do something similar to what we do for `Image`, and treat a string without a protocol as a file (although not relative to the classpath as in the case of images), turning it into a `file:` URL internally where needed.

@prrace what do you think?

Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: > As a potential user of this API I have to ask, why would this be a URL instead of a File or Path ? Particularly if only "file:" URLs will be valid. I can't think of a scenario where it won't always be an extra inconvenience to convert the path I have to a URL.

I will think about it but not promising yet.
 We only forsee (for now) a file ..  We use URL a lot and also the javax.print API in 2D uses URL .. and so 2D hands us a file URL as well as expecting one .. and FX printing depends heavily on that right now so there's no escaping URLs somewhere.

Comment (johanvos) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: I agree with @kevinrushforth that if a String without protocol is passed, it should be treated as a file (absolute or relative to <what>?) 
I'm also not sure that the URL should be exposed here. I understand it's needed in the lower-level print API but you already do the conversion in the `syncOutputFile` method. Hence, since only the file protocol is supported, it might be easier for API users to pass the location of the file instead of a URL.
In case later other URL protocols are supported, a property `outputURL` might be introduced?

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:474]: I like the flexibility and consistency of defining it as a URL, as long as we also interpret a url without a scheme as a file name. Borrowing language from the Image docs, perhaps something like this?

```
The URL string can either be a URL with a "file:" protocol that can be resolved
by @link java.net.URL} or a file path that can be resolved by {@link java.io.File}.
```

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 6 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:541]: It's still a good idea to follow the pattern, so yes let's make the new methods final. We can file a cleanup bug for the existing ones (and since the class is final, we can do it without compatibility concerns).

Also, I built the docs, and it is better to not add any javadoc comments on the setter and getter.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 7]
```diff
@@ -462,6 +462,97 @@ public void setJobName(String name) {
...
+     * <p>
+     * If the application does not have permission to write to the specified
+     * file, a {@code SecurityException} may be thrown when printing.
+     * If the print system does not support print-to-file, then this
+     * setting will be ignored.
+     * <p>
+     * If the URL specifies a non-existent path, or does not specify
+     * a writable file it may be ignored, or a SecurityException may be thrown.
+     * The default value is an empty string, which is interpreted as unset,
+     * which means output is sent to the printer.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:492]: Can you add a `* @defaultValue empty string` javadoc tag?

**Review by kevinrushforth — COMMENTED** [review]
Overall the new API and functionality looks good, and fills a void that was discussed [in this thread](https://mail.openjdk.java.net/pipermail/openjfx-dev/2019-May/023292.html) a couple years ago.

General comments:

* If the user selects "print to file" from the native dialog, it will write to the `outputFile` property. This should be documented.

* On the question of error handling. There are two types of errors we could consider. 

The first, which is the one we had been discussing, is where the URL is a valid file URL, but the program can't write to it, either because we are running with a security manager and the application doesn't have permission, or because we can't write to the output file (e.g., the path doesn't exist or is read-only).

The second is a bad URL. If the URL is malformed or uses an unsupported protocol (anything other than `file:`), we could throw IllegalArgumentException, either when the invalid URL is set or when the print job is run. We might want to do this regardless of what we do about the first type of error. This could be done as simple as adding `@throws IllegalArgumentException if the url uses a protocol other than "file:"`. What do you think?

For files that can't be written (the first type of error), we have three choices:

1. Define it loosely such that using a file that cannot be written to will either throw an exception when the print job is run or it will be ignored.

2. Define it to always throw an exception. We would need to add validation logic in PrintJob before calling the implementation methods, but we wouldn't necessarily be able to catch all errors (e.g., an IOEXception that only happens after you start writing to the file). This seems like overkill.

3. Define it to never throw an exception. The implementation would need to wrap the call to print in a try / catch, probably logging the result.

I would probably lean towards 1 (which is what you are suggesting), but 3 could work as well.

I left a couple inline comments on the API.

File: modules/javafx.graphics/src/main/java/com/sun/prism/j2d/print/J2DPrinterJob.java [hunk 8]
```diff
@@ -797,6 +835,13 @@ private void checkPermissions() {
         SecurityManager security = System.getSecurityManager();
         if (security != null) {
             security.checkPrintJobAccess();
+            String file = settings.getOutputFile();
+            if (!file.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/prism/j2d/print/J2DPrinterJob.java:839]: The default value for the property is the empty string. But it does bring up a good point that we should either check and throw NPE if `setOutputFile` is called with `null` or we should map null to the empty string.

File: tests/manual/printing/PrintToFileTest.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            }
+
+            Platform.runLater(() -> {
+                Text t = new Text("file="+settings.getOutputFile());
+                root.getChildren().add(t);
+            });
+            Text printNode = new Text("\n\nTEST\nabc\ndef");
+            job.printPage(printNode);
+            job.endJob();
+            if (f.exists()) {
```
Comment (kevinrushforth) [inline tests/manual/printing/PrintToFileTest.java:134]: I have only tried this on Windows. This check is failing for me when I run the test, and yet the file is created correctly. If I add a short sleep, it works (so the printing is likely being done on a different thread).

File: tests/manual/printing/PrintToFileTest.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                System.out.println("created file");
+                passed = true;
+                f.delete();
```
Comment (kevinrushforth) [inline tests/manual/printing/PrintToFileTest.java:137]: Maybe leave the file? I found it useful to verify its contents.

**Review by kevinrushforth — COMMENTED** [review]
I left a comment on the implementation and a couple on the test inline.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 6 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:546]: I missed the part where you said you would file the bug, so I filed [JDK-8269638](https://bugs.openjdk.java.net/browse/JDK-8269638).

**Review by kevinrushforth — COMMENTED** [review]
The API looks fine now, although I may have some minor wording suggestions after I do another pass tomorrow. Go ahead and file the CSR and move it to proposed as soon as you can to see if Joe has any comments in parallel with our finishing the review.

I also need to do some testing on Mac and Linux (the manual test now works as expected on Windows).

**Comment by prrace** [comment]
CSR: https://bugs.openjdk.java.net/browse/JDK-8269642

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 11]
```diff
@@ -462,6 +462,98 @@ public void setJobName(String name) {
...
+     * driver should spool the rendered print data.
+     * <p>
+     * Applications can use this to programmatically request print-to-file
+     * behavior where the native print system is capable of spooling the
+     * output to a filesystem file, rather than the printer device.
+     * <p>
+     * This is often useful where the printer driver generates a format
+     * such as Postscript or PDF, and the application intends to distribute
+     * the result instead of printing it, or for some other reason the
+     * application does not want physical media (paper) emitted by the printer.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:481]: Very minor: maybe consider combining the first three paragraphs into a single paragraph?

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 12]
```diff
@@ -462,6 +462,98 @@ public void setJobName(String name) {
...
+     * <p>
+     * The default value is an empty string, which is interpreted as unset,
+     * equivalent to null, which means output is sent to the printer.
+     * So in order to reset to print to the printer, clear the value of
+     * this property by setting it to null or an empty string.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:486]: This doesn't flow as well as it could. I think you only need to mention once that `null` is the same as an empty string, and then you can just say "empty string". Maybe something like this?

```
The default value is an empty string, which means output is sent to the printer. So in order to reset
to print to the printer, clear the value of this property by setting it to an empty string. A value
of {@code null} is treated as an empty string.
```

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 13]
```diff
@@ -462,6 +462,98 @@ public void setJobName(String name) {
...
+     * <p>
+     * Additionally if the application displays a printer dialog which allows
+     * the user to specify a file destination including altering an application
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:489]: Minor: I think there should be a comma between `destination` and `including`.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 14]
```diff
@@ -462,6 +462,98 @@ public void setJobName(String name) {
...
+     * specified file destination, the value of this property will reflect that
+     * user-specified choice, including clearing it to re-set to print to
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:491]: `reset` is one word (no need to hyphenate).

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java [hunk 15]
```diff
@@ -462,6 +462,98 @@ public void setJobName(String name) {
...
+     * the printer, if the user does so.
+     * <p>
+     * If the print system does not support print-to-file, then this
+     * setting will be ignored.
+     * <p>
+     * If the specified name specifies a non-existent path, or does not specify
+     * a user writable file, when printing the results are platform-dependent, including
+     * replacement with a default output file location, printing to the printer instead,
+     * or a platform printing error.
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:500]: This sentence is a bit awkward and hard to parse. Maybe break it into two sentences? Perhaps something like this:

```
If the specified name specifies a non-existent path, or does not specify a user writable file, the
results when printing are platform-dependent. Possible behavior might include replacement with
a default output file location, printing to the printer instead, or a platform printing error.
```

**Review by kevinrushforth — COMMENTED** [review]
I've added my final comments on the API docs. Some of them are pretty nit-picky, so take them as suggestions.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 11 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:481]: well .. I usually like a short paragraph that succinctly says what it does as the first paragraph

Anyway I've re-read all this and I prefer it as it is.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 12 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:486]: But I don't say it twice. I say
>  The default value is an empty string, which is interpreted as unset, equivalent to null,

and 
>  clear the value of this property by setting it to null or an empty string.

which is somewhat different and makes i t very clear that either will work .. 

So I think this is all fine and flows as intended.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 13 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java]: yep

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 14 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java]: it is reset elsewhere not sure why a hyphen is here.

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 15 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java]: ok split it

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 11 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:481]: ok

File: modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java (hunk 12 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/print/JobSettings.java:486]: ok

**Review by kevinrushforth — APPROVED** [review]
Looks good.

**Review by prsadhuk — APPROVED** [review]
